L1Trigger/Phase2L1ParticleFlow/plugins/L1TPFProducer.cc

```diff
@@ -271,13 +279,19 @@ L1TPFProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
         l1tpf_impl::writeManyToFile(l1regions_.regions(), fRegionDump);
     }
 
-    // Then save the regions to the COE file
+    // Then save the regions to the COE and/or APx pattern file(s)
     // Do it here because there is some sorting going on in a later function
-    if (fRegionCOE && fRegionCOE->is_open() && neventsproduced<neventscoemax) {
-        std::vector<l1tpf_impl::Region> regions = l1regions_.regions();
-        fRegionCOE->writeTracksToFile(regions,neventsproduced==0);
+    std::vector<l1tpf_impl::Region> regions = l1regions_.regions();
+    if (fRegionCOE) {
+        if (!fRegionCOE->is_full()) fRegionCOE->storeTracks(regions);
+        if (fRegionCOE->is_full() && nEventsProduced==nEventsCOEMax-1) fRegionCOE->writeObjectsToFiles();
```

Comment:
this could lead to cases where the file is never writte because the dump gets filled too early or too late.
since the pattern file has already some logic inside it for counting the numbers of received frames etc, maybe all this could become a `if (fRegionCOE) fRegionCOE->storeTracks(l1regions_.regions());`  
That would make it safer, shorter (especially since now things get duplicated for COE and APx), and you also avoid making a full copy by value of the regions (which you were doing in line 284 for no obvious reason)